DRAFT: Adding KEP: Preemption Toleration

everpeace opened the pull request.

everpeace committed:
Adding KEP: Preemption Toleration

shioshiota reviewed (commented):
If `minimumRunningDuration` prevents preemption of the pod by pods of any priority class pods, even important pods, including pod for managing cluster, cannot preempt it.

How about changing `minimumRunningDuration` to  `expireDuration`?

### example1

```
kind: PriorityClass
metadata:
  name: low-1h
value: 8000
```
```
minimumPreemptablePriority: 10000
expireDuration: 1h
```
This means only pods with higher priority than 10000 can preempt the pods of low-1h pods if the pod doesn't for 1 hour.
After 1 hour of the pod starts, any pods that have higher priority can preempt it.

### example2

```
kind: PriorityClass
metadata:
  name:  low-non-preempted
value: 8000
```

```
minimumPreemptablePriority: 10000
```
This means only pods that have higher priority than 10000 can preempt the pods of low-non-preempted pods. `minimumPreemptablePriority` will not expire.

ordovicia commented on kep/NNN-preemption-toleration/README.md:
```diff
@@ -0,0 +1,125 @@
+
+So, cluster administrators would like to exempt pods that do not finish the first epoch from preemption even if the job has lower priority.
+
+However, it is hard to monitor when the first epoch is finished outside of the job. Thus, a minimum running time guarantee is useful practically.  For example, cluster administrator would introduce several priority classes with the same priority value for introducing several minimum running time guarantee categories like below:
```
Does the "running time" include pods' initialization phase (pulling images, running init containers, etc) or, only include duration after pods reach the `Running` phase?
We should determine an exact definition and avoid the term "minimumRunningDuration" because it could be confusing.

everpeace replied:
I think we could use time from being scheduled.

ordovicia commented on kep/NNN-preemption-toleration/README.md:
```diff
@@ -0,0 +1,125 @@
+type PreemptionToleration struct {
+	metav1.TypeMeta `json:",inline"`
+
+  // MinimumPreemptablePriority specifies the minimum priority value that can preempt this priority class.
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // MinimumRunningDuration specified the minimum running duration of pods that can be preempted.
+  MinimumRunningDuration *time.Duration `json:"minimumRunningDuration,omitempty"`
```
Are these two conditions OR'e or AND'ed?

everpeace replied:
I thought it should be AND'ed.

everpeace commented:
> This means only pods with higher priority than 10000 can preempt the pods of low-1h pods if the pod doesn't for 1 hour.
> After 1 hour of the pod starts, any pods that have higher priority can preempt it.

👍  

> How about changing `minimumRunningDuration` to `expireDuration`?

How about using `toleraitonSeconds` as similar to pod toleration terminology?  It's more familiar to more users.

shioshiota commented:
> How about using toleraitonSeconds as similar to pod toleration terminology? It's more familiar to more users.

👍

shioshiota reviewed (approved):
LGTM 👍

ordovicia commented on kep/NNN-preemption-toleration/README.md:
```diff
@@ -0,0 +1,126 @@
+  MinimumPreemptablePriority *int32 `json:"minimumPreemptablePriority,omitempty"`
+
+  // TolerationSeconds specified how long this priority class can tolerate preemption by priorities lower than MinimumPreemptablePriority.  Null value specifies forever (default).  Duration means the duration from the pod being scheduled to some node.
+  TolerationSeconds *time.Duration `json:"tolerationSeconds,omitempty"`
```
`TolerationSeconds` should have an integer type because it is "seconds", not "duration".
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#toleration-v1-core

everpeace replied:
great catch!  I also noticed this last night and will update the PR soon 👍

ordovicia reviewed (approved):
LGTM except a small comment.

everpeace committed:
add explicit description that "tolerationSeconds" does affect only to scheduled pods

everpeace committed:
fix unit of "tolerationSeconds"

everpeace commented:
@shioshiota @ordovicia I updated the branch on the two points:

- added explicit explanation "tolerationSeconds" only affects to _scheduled_ pods (except for nominated pods https://github.com/everpeace/scheduler-plugins/pull/1/commits/7e7db2f83966e88243ea850ebcf449992203f668
- fix unit of "tolerationSeconds" with `*int64` https://github.com/everpeace/scheduler-plugins/pull/1/commits/ac0c1e505948342f28d8553d328ee13ec9114259

PTAL 🙇

everpeace commented:
@shioshiota I'm gonna close the PR and will create a squashed branch for upstream.
